### PR TITLE
fix(developer): omnibus package of fixes for kmc 🗜

### LINKED ISSUE #9154: chore(developer): investigate build time for vietnamese_ keyboards 🗜
> Note: building this keyboard is very slow. Some time complexity clearly involved (plus impressively inefficient duplication of constructed static strings in `if` statements in the keyboard)

This needs separate investigation.

### LINKED ISSUE #9156: bug(developer): kmc/kmw needs to strip codes from store outputs 🗜
kmc/kmw-compiler needs to strip the UC_SENTINEL + CODE_x values from strings it is emitting, replacing with '.', as was done in the legacy compiler.

output\17.0.107-alpha-local\aksarabali_panlex.js is not equal to output\kmc\aksarabali_panlex.js

This is correct:
```js
  this.s_cons_key_22="..........................................";
  this.s_cons_main_23="ᭉᬯᬭᬢᬬᬧᬲᬤᬖᬕᬳᬚᬓᬮᬜᬜᬡᬘᬪᬩᬦᬫᬱᬰᬗᬨᬝᬞᬣᬟᬠᬥᬛᬔᬙᭆᭅᭈᭉᭇᭊᭋ";
```

kmc is generating:
```js
  this.s_cons_key_22=".\n䀀Q\x10.\n䀀W\x10.\n䀀R\x10.\n䀀T\x10.\n䀀Y\x10.\n䀀P\x10.\n䀀S\x10.\n䀀D\x10.\n䀀F\x10.\n䀀G\x10.\n䀀H\x10.\n䀀J\x10.\n䀀K\x10.\n䀀L\x10.\n䀀Z\x10.\n䀐Z\x10.\n䀀X\x10.\n䀀C\x10.\n䀀V\x10.\n䀀B\x10.\n䀀N\x10.\n䀀M\x10.\n䀀Û\x10.\n䀀Ý\x10.\n䀀Ü\x10.\n䀐Ü\x10.\n䀀À\x10.\n䀀Ā\x10.\n䀀ā\x10.\n䀀Ă\x10.\n䀀ă\x10.\n䀀Ą\x10.\n䀀ą\x10.\n䀀Ć\x10.\n䀀ć\x10.\n䀀Ĉ\x10.\n䀀ĉ\x10.\n䀀Ċ\x10.\n䀀ċ\x10.\n䀀Č\x10.\n䀀č\x10.\n䀀Ď\x10";
  this.s_cons_main_23="ᭉᬯᬭᬢᬬᬧᬲᬤᬖᬕᬳᬚᬓᬮᬜᬜᬡᬘᬪᬩᬦᬫᬱᬰᬗᬨᬝᬞᬣᬟᬠᬥᬛᬔᬙᭆᭅᭈᭉᭇᭊᭋ";
```

### LINKED ISSUE #9157: bug(developer): kmc/kmw needs to strip \r from HTML 🗜
My guess is that `\r` needs to be stripped from the embedded HTML, rather than be converted to space, in `KH` in order to match the output from kmcomp.

For keyboard release/a/anii:

kmc is emitting:
```
  this.KH='<!DOCTYPE html>  <html lang="en" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml">  <head>  ...
```

kmcomp emitted:
```
  this.KH='<!DOCTYPE html> <html lang="en" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml"> <head> ...
```

### LINKED ISSUE #9159: bug(developer): kmc-kmn/kmw needs to cleanup bitmap VK keys before emitting 🗜
This impacted release/b/balochi_inpage.

### LINKED ISSUE #9160: bug(developer): kmc-kmn/kmw should add/delete certain fields in touch layout when writing 🗜
* The `platform.displayUnderlying` flag is always emitted by kmcomp, so kmc should always emit it too.
* `platform.font` and `platform.fontsize` should be eliminated if empty string
* `key.id` should be eliminated if empty string

This was encountered in release/a/aksarabali_panlex.

### LINKED ISSUE #9161: bug(developer): kmc-kmn/kmw needs to convert all `*ZWNJ*` to `<|>`  🗜
For release/b/balochi_inpage:

```
JSON not equal phone>layer>1>row>3>key>3>text>(value): <|> !== *ZWNJ*
ERROR: javascript touch layout is not equal
ERROR: output\17.0.107-alpha-local\balochi_inpage.js is not equal to output\kmc\balochi_inpage.js
```

It looks like the conversion is matching the first result but not subsequent results.

### LINKED ISSUE #9177: bug(developer): vk constants requires `K_oDF` and `K_?C1` 🗜
Discovered in release/basic/basic_kbdiulat (`K_oDF`) and release/basic/basic_kbdbr, these codes are used by some legacy keyboards, although they are not surfaced in the designer.

### LINKED ISSUE #9178: bug(developer): touch layout default values for `width`, `hint`, and `pad` should not be emitted in compiled layout 🗜
The touch layout writer should strip these values out if found:

* `width: 0` (as found in release/sil/sil_devanagari)
* `pad: 0` (as found in release/basic/basic_kbdinasa)
* `hint: ""` (as found in experimental/c/carian)

(Just listing the first keyboard found in each case.)

### LINKED ISSUE #9180: bug(developer): kmc does not support `bitmap` field in visual keyboards 🗜
A number of keyboards include bitmap keys in their on screen keyboards, for example release/b/balochi_inpage. We need to support reading and writing the bitmaps in kvk and kvks.

### LINKED ISSUE #9181: bug(developer): kmc - call statements were wrongly referenced and indexed 🗜
In release/k/korean_rr, the output from kmc did not match kmcomp because:

1. The call statements were incorrectly indexed on first use
2. The .call_js files were misreferenced.

### LINKED ISSUE #9182: bug(developer): kmc - font names were incorrectly escaped in compiled kmw visual keyboard 🗜
Quote characters were not escaped in, for example, release/sil/sil_nko.

### LINKED ISSUE #9183: bug(developer): whitespace handling is not 100% in imported files in kmw compiler 🗜
Whitespace handling was not 100% identical to kmcomp for a number of imports:

* CSS: blank line required at EOF if not present, e.g. release/sil/sil_akha_act, but only if the CSS file was not empty (release/n/nailangs)!
* Embedded HTML needs a blank line at EOF if not present, e.g. release/n/nasa_yuwe
* `\r` is converted to `\\n` and `\n` is converted to ` `, to match kmcomp (e.g. release/sil/sil_akha_act)

### LINKED ISSUE #9184: bug(developer): kvks reader needs to strip pure whitespace for key text 🗜
For example, in release/b/bangla_munir. This is what legacy kmcomp did.

### LINKED ISSUE #9185: bug(developer): off-by-one in some index calculations for outputting js in kmc/kmw-compiler 🗜
This was a little complex, because moving from Delphi, which has 1-based string indices, to Javascript, which has 0-based indices, and trying to map that to Keyman's 1-based index model, is fiddly.

For example in release/a/armenian_mnemonic.

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #9154.
Fixes #9156.
Fixes #9157.
Fixes #9159.
Fixes #9160.
Fixes #9161.
Fixes #9177.
Fixes #9178.
Fixes #9180.
Fixes #9181.
Fixes #9182.
Fixes #9183.
Fixes #9184.
Fixes #9185.

This can be reviewd commit-by-commit to see each fix separately. Combined into one PR to reduce clutter. This, together with some small patches to the keyboards repository, means that the kmw compiler and kvk compiler now have a 100% match with kmcomp. Some of the fixes are inconsequential, such as whitespace matching, but most of the others are important.

One of the fixes does not change behaviour but has a dramatic performance improvement (reducing vietnamese_telex build time from multiple minutes to 6 seconds).

The following fixes are in this PR:

* fix(developer): strip pure whitespace from key text in kvks reader (#9184)

   kmcomp was unable to track pure whitespace in .kvks files. To match this behaviour, we should strip pure space whitespace from key element values in .kvks files only.

* fix(developer): off-by-one index calculations in js output (#9185)

* fix(developer): kmc: cleanup handling of whitespace in import files (#9183)

   * CSS: blank line required at EOF if not present, but only if the CSS file was not empty!
   * Embedded HTML needs a blank line at EOF if not present
   * `\r` is converted to `\\n` and `\n` is converted to ` `, to match kmcomp

* fix(developer): escape font name correctly in vk (#9182)

* fix(developer): `call` statements were misreferenced in kmc (#9181)

   * First use of a `call` statement had an incorrect index (misuse of `push` return value)
   * The .call_js files were incorrectly prepended with the basename of the keyboard

* fix(developer): kmc - support `bitmap` field in .kvk/kvks files (#9180)

* fix(developer): strip default values when compiling touch layout (#9178)

   The values `pad`, `width`, and `hint` should not be emitted in a compiled touch layout file, if they have default values of `0`, `0`, or `""`, respectively. This matches the legacy kmcomp pattern.

* fix(developer): vk constants requires `K_oDF` and `K_?C1` (#9177)

   Some imported visual keyboards include `K_oDF` or `K_?C1` key codes. These are not made visible in the UI but as they are "valid" VK codes, they should not cause an error.

* fix(developer): kmc/kmw was not stripping codes correctly (#9156)

   The compiler was only skipping the first codepoint in a CODE sequence.

* fix(developer): kvks file reader needs to handle bitmaps (#9159)

   If a key contains a bitmap, then ignore the text for the key.

* fix(developer): touch layout was not converting all special keycaps (#9161)

   kmw-compiler was converting only the first key cap with `*special*` caption, rather than all key caps.

* fix(developer): improve kmcmplib/kmw compiler performance (#9154)

   Improves performance in both kmcmplib and in kmc-kmn/kmw-compiler. After these changes are applied, vietnamese_telex keyboard builds in around 6 seconds on my machine (down from over 2 minutes). This is certainly only the surface of performance improvements we could make.

   Addresses excessive memory reallocations in kmcmplib, by reallocating store and key arrays in large chunks of 100 items rather than reallocating on each new item added. This resulted in a 250x speed-up on functions such as `AddStore` in WASM build.

   Some careful modulus arithmetic was needed for resizing the key array, which can be grown in larger increments.

   Resolved excessive zeroing out of destination buffer in u16ncpy, which was being called with an 8kb destination buffer on every line of the file.

   In the kmw side, vast majority of the performance cost was in copying of buffers while loading a .kmx into memory, rather than creating views into the memory. Essentially replaced this pattern:

   ```js
   return this.rString.fromBuffer(source.slice(offset));
   ```

   with:

   ```js
   const data = new Uint8Array(source.buffer, source.byteOffset + offset);
   return this.rString.fromBuffer(data);
   ```

* fix(developer): kmc-kmn/kmw should strip `\r` from inline html (#9157)

   Matching kmcomp behaviour once more.

* fix(developer): kmc-kmn/kmw should add/delete certain fields in touch layout when writing (#9160)

   To match the behaviour of kmcomp, kmc should add/delete the following fields when writing out a kmw touch layout:
   * `platform.displayUnderlying` flag is always emitted by kmcomp, so kmc should always emit it too.
   * `platform.font` and `platform.fontsize` should be eliminated if empty string
   * `key.id` should be eliminated if empty string

@keymanapp-test-bot skip

### CODE REVIEW

File: developer/src/kmc-kmn/src/kmw-compiler/kmw-compiler.ts [hunk 1]
```diff
@@ -140,6 +144,7 @@ export function WriteCompiledKeyboard(
     try {
       const data = callbacks.loadFile(sEmbedCSSFilename);
       sEmbedCSS = new TextDecoder().decode(data);
+      if(sEmbedCSS != '' && !sEmbedCSS.endsWith('\r\n')) sEmbedCSS += '\r\n';  // match CompileKeymanWeb.pas
```
Comment (jahorton) [inline developer/src/kmc-kmn/src/kmw-compiler/kmw-compiler.ts:147]: Wait... so it actually does emit `\r`... for CSS files, despite erasing / replacing them for HTML?

File: common/web/types/src/kvk/kvks-file-reader.ts [hunk 2]
```diff
@@ -131,7 +131,7 @@ export default class KVKSFileReader {
           let key: VisualKeyboardKey = {
             flags: isUnicode ? VisualKeyboardKeyFlags.kvkkUnicode : 0, // TODO-LDML: bitmap support
             shift: shift,
-            text: sourceKey._ ?? '',
+            text: sourceKey.bitmap ? '' : (sourceKey._ ?? ''),
```
Comment (jahorton) [inline common/web/types/src/kvk/kvks-file-reader.ts:142]: Kind of curious what this is about.  To my knowledge, KMW doesn't have any knowledge of per-key "bitmaps" - be that image-bitmaps or bit masks (to choose a distinct word).

**Review by jahorton — APPROVED** [review]
LGTM, I guess?  Still wondering what the visual-keyboard layout "bitmap" thing (for certain keys?) being referenced in a few commits is, though.

File: developer/src/kmc-kmn/src/kmw-compiler/kmw-compiler.ts (hunk 1 above)
Comment (mcdurdin) [reply developer/src/kmc-kmn/src/kmw-compiler/kmw-compiler.ts:147]: The `\r\n` is replaced by the `RequotedString` function later... So I could have just added a space but this was a closer match to the existing code.

File: common/web/types/src/kvk/kvks-file-reader.ts (hunk 2 above)
Comment (mcdurdin) [reply common/web/types/src/kvk/kvks-file-reader.ts:142]: Yes, that is right, KMW does not have support for per-key bitmaps. But .kvk and .kvks file formats do support per-key bitmaps, and these are used in Keyman for Windows.

**Pull request merged**